### DISCUSSION TITLE
Multiple Mountpoints (Multiple LIQ files or a single LIQ file?)

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#1702] [Q&A] [answered]
Can we have multiple LIQ files? (Multiple mountpoints)
Do we have to update the Icecast.xml file with each of the new Mountpoints?

Would it be best to have a single LIQ file with all mountpoints listed within it?
Or, would it be best to have one for each DJ/Station?

What if you have 1,000's of DJ/Stations (Mountpoints).
Will the icecast.xml file get too big, or is it made not to affect its performance when many mountpoints are added?

One more thing.
If we add in a new Mountpoint, rather a new LIQ or within the same LIQ.
We have to restart Liquidsoap. Do we also have to restart Icecast as well?
And will this restarting of either of them affect the already running Stations, and cause them to start over again?

Thanks.
Wayne

**Comment by oranjebloom** [comment]
Why do you need multiple mount points? Is each mount point for different streaming content?
I have two liq files. I have one icecast.xml that lists the mount points for those two liq files.
Each mount point streams different content. This works flawlessly for me.

**Comment by oranjebloom** [comment]
You don't need to restart icecast if you want to restart the liq scripts. And yes, when you restart the liq scripts, you will end the current streams, kick all listeners and start from scratch.

**Reply by carrzkiss** [reply]
Let's say I have 50 stations.
Each station has its own playlist.
And each station comes online at a certain time.
(or)
I add in a new station to the list to make it 51 stations.

Is there a way to start/restart only that station?

Also, is there a better way of loading the stations?
Example:
(In ASP, or any Server-Side language, would do something like this for dynamic content.
ID-GOES-HERE = request.QueryString("ID")
And then
Stations  = playlist("https://domain.com/DJ/Station.asp?ID=<%=ID-GOES-HERE%>",mode="normal")
And this will load the playlist for that particular station.
So, if the URL is
Station.asp?ID=1
Station.asp?ID=2
Station.asp?ID=3
Then it will load it of these stations depending on its ID#

**Comment by toots** [comment]
The general trade off is:
* Listing all mount points in a single `liq` files makes it easier to track/restart services.
* Using multiple liq file makes it easier to add/update/remove mountpoints.
I'm not sure what you mean with icecast.xml? In general, icecast is never restarted.

**Reply by carrzkiss** [reply]
Sorry that I have not responded to either of you. I have had other life issues to tend with here lately, which has kept me right busy.

Let's say I have stations added daily for customers.

1. If I add them all into a single liq file, I will have to wait until a scheduled maintenance time to start the new station for the customer since it is inside an existing liq file, Say, 12:00 midnight, which will only take a few seconds to do the restart, and all DJ's will know about this time, so they will not lose any listeners.
2. Using multiple LIQ files.
How would I do it like this, as that is something more similar to what I would want to do?
And since you stated I would be able to "add/update/remove mountpoints." 
Would I also be able to "start | restart" each station, or would they all still have to be restarted together?

Thanks.
Wayne

**Comment by carrzkiss** [comment, marked as answer]
Figured out how to run multiple Stations and start and stop each one without messing with the other stations.
Love it.
Working great.

Put all scripts in the [Script] folder.
Run the following line.
./daemonize-liquidsoap.sh 101.liq
Then to start that stream.
sudo systemctl start 101.liq-liquidsoap
sudo systemctl status 101.liq-liquidsoap

If you have any issues with the above, please look at the following article to resolve issues you might face.
[CFF Knowledge Base Article 3236- Unable to find liquidsoap_binary in your path](https://www.cffkb.com/Main.asp?irid=3236&Type=Article)

Thanks to **oranjebloom** and **toots** for helping me get this far.
Now to move on to the next step.
Setting up a Load Balancer, SSL, and then set the system up to allow live updates to the playlist.
Love learning this and having all this information to look back on.

Thanks, everyone,
Wayne

**Comment by carrzkiss** [comment]
@toots @oranjebloom 
I lost all my notes for Installing Liquidsoap properly, and no longer have access to the "Slack" chat logs I shared with oranjebloom
And was wondering if either of you could send me over the chat logs for that session?
Thank you.
Wayne

**Reply by oranjebloom** [reply]
hi @carrzkiss 
I can't seem to find that chat either. I believe Slack doesn't allow for archiving of chats (from what I understand). But if you are needing to install liquidsoap version 2.0.2 via opam on Debian, then you can try the instructions below (there is actually a newer version 2.0.3 but I would wait a little longer as there may be some bugs getting ironed out with that release).

```
opam switch create 2.0.2 4.13.1

opam init
eval $(opam env)
opam install depext
```

If you want to use mad then go with this Installation
```
opam depext taglib mad lame cry ssl samplerate magic lastfm inotify ocurl liquidsoap
opam install taglib mad lame cry ssl samplerate magic lastfm inotify ocurl liquidsoap

```
... but I am using ffmpeg these days as it is the recommended way >> ffmpeg Installation
```
opam depext taglib ffmpeg lame cry ssl samplerate magic lastfm inotify ocurl liquidsoap
opam install taglib ffmpeg lame cry ssl samplerate magic lastfm inotify ocurl liquidsoap
```

**Reply by oranjebloom** [reply]
I believe that the daemon has also been updated recently, so you should install the most recent daemon. I would then follow the notes above about daemonising the .liq scripts. I remember that part was a little tricky for you.

**Reply by carrzkiss** [reply]
I remember I had to add permissions for specific folders. 
FOUND IT!
I forgot that I had written it up on StackOverflow.

Once you create your user, "Icedemo."
(logged in as root)

`adduser Icedemo` 
Give user SUDO rights.

`usermod -aG sudo Icedemo`
Now give permissions to Icedemo to all the folders.

```
/opt/liquidsoap/Playlists/
/var/log/liquidsoap/
/etc/icecast2/web/
/etc/icecast2/admin/
```
Use the following script to give permissions.

```
cd /opt
sudo chown -Rv icedemo:icedemo liquidsoap/
cd /var/log/
sudo chown -Rv icedemo:icedemo liquidsoap/
cd /etc/icecast2/
sudo chown -Rv icedemo:icedemo web/
sudo chown -Rv icedemo:icedemo admin/
```
Leave permission the same for the icecast.xml file.

Log in as your user Icedemo

`su Icedemo`  
Once you have done all the above and your script is formatted correctly - something like this.

```
# !/home/iceadmin/.opam/default/bin/liquidsoap
# log dir
set ("log.file.path","/var/log/liquidsoap/basic-radio.log")
set("scheduler.fast_queues",1)
 # Music
music = playlist("/opt/liquidsoap/playlists/RadioOne.m3u",mode="normal",reload_mode="watch")

# Start building the feed with music
radio = music

# Stream it out
output.icecast(%mp3(bitrate=256,samplerate=44100,internal_quality=0,id3v2=true,stereo=true,stereo_mode="stereo"),
name="Radio",
encoding="UTF-8",
host="192.168.2.203",
port=8000,
password="hackme",
icy_metadata="true",
description="Internet Radio Station",
mount="Radio",
mksafe(radio))
```
We will need to restart each service.

```
#Restart IceCast   
sudo systemctl restart icecast2
#Check status of IceCast  
sudo systemctl status icecast2
(Ctrl+Z to go to prompt)
```
Next, we will need to do the same to Liquidsoap.

```
#Restart Liquidsoap
sudo systemctl restart radio.liq-liquidsoap
#Check status of  Liquidsoap
sudo systemctl status radio.liq-liquidsoap
#(After checking the status, you can use (Ctrl+Z to go to prompt))
```
Your output of the Liquidsoap status should look similar to mine.

```
Icedemo@linuxcore-03:~/liquidsoap-daemon$ sudo systemctl status radio.liq-liquidsoap
● radio.liq-liquidsoap.service - radio.liq Liquidsoap daemon
     Loaded: loaded (/etc/systemd/system/radio.liq-liquidsoap.service; disabled; vendor preset: enabled)
     Active: active (running) since Mon 2021-05-24 03:47:01 UTC; 4s ago
       Docs: http://liquidsoap.fm/
    Process: 54678 ExecStart=/home/Icedemo/.opam/default/bin/liquidsoap /home/Icedemo/liquidsoap-daemon/script/radio.liq-run.liq (code=exited, status=0/SUCCESS)
   Main PID: 54680 (liquidsoap)
      Tasks: 9 (limit: 2251)
     Memory: 17.3M
     CGroup: /system.slice/radio.liq-liquidsoap.service
             └─54680 /home/Icedemo/.opam/default/bin/liquidsoap /home/Icedemo/liquidsoap-daemon/script/radio.liq-run.liq
May 24 03:47:01 linuxcore-03 systemd[1]: Starting radio.liq Liquidsoap daemon...
May 24 03:47:01 linuxcore-03 systemd[1]: Started radio.liq Liquidsoap daemon.
```
If the above is showing: Active: active (running)
Next, check your Icecast webpage, and you should see your mountpoints, and your radio should be working.
To access the radio station in our example script above would be.

`(local) 192.168.2.203:8000/Radio `
This will load the player in its own browser window, and your stream should start on its own. If not, click the PLAY button.

That's it; that is how you get your streaming radio to work.
I hope this helps others out.

Wayne
